src/main/java/org/folio/inventory/consortium/consumers/ConsortiumInstanceSharingHandler.java

```diff
@@ -0,0 +1,277 @@
+  private final KafkaConfig kafkaConfig;
+  private final Map<String, KafkaProducer<String, String>> producerList = new HashMap<>();
+  protected static final ObjectMapper OBJECT_MAPPER = new ObjectMapper().setSerializationInclusion(JsonInclude.Include.NON_NULL)
+    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    .configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true);
+
+  public ConsortiumInstanceSharingHandler(Vertx vertx, Storage storage, KafkaConfig kafkaConfig) {
+    this.vertx = vertx;
+    this.storage = storage;
+    this.kafkaConfig = kafkaConfig;
+  }
+
+  @Override
+  public Future<String> handle(KafkaConsumerRecord<String, String> record) {
+    try {
+      Promise<String> promise = Promise.promise();
+      String consortiumId = record.key();
+      LOGGER.info("handle :: CONSORTIUM_INSTANCE_SHARING_INIT from consortiumId {}", consortiumId);
+      SharingInstance sharingInstance = Json.decodeValue(record.value(), SharingInstance.class);
+
+      String instanceId = sharingInstance.getInstanceIdentifier().toString();
+      Map<String, String> kafkaHeaders = KafkaHeaderUtils.kafkaHeadersToMap(record.headers());
+
+      LOGGER.info("Event CONSORTIUM_INSTANCE_SHARING_INIT has been received for instanceId: {}, sourceTenant: {}, targetTenant: {}",
+        instanceId, sharingInstance.getSourceTenantId(), sharingInstance.getTargetTenantId());
+      String tenantId = kafkaHeaders.get(OKAPI_TENANT_HEADER);
+
+      Context targetTenantContext = EventHandlingUtil.constructContext(sharingInstance.getTargetTenantId(),
+        kafkaHeaders.get(OKAPI_TOKEN_HEADER), kafkaHeaders.get(OKAPI_URL_HEADER));
+      InstanceCollection targetInstanceCollection = storage.getInstanceCollection(targetTenantContext);
+
+      Context sourceTenantContext = EventHandlingUtil.constructContext(sharingInstance.getSourceTenantId(),
+        kafkaHeaders.get(OKAPI_TOKEN_HEADER), kafkaHeaders.get(OKAPI_URL_HEADER));
+      InstanceCollection sourceInstanceCollection = storage.getInstanceCollection(sourceTenantContext);
+
+      LOGGER.info("handle :: checking is instance {} exists on tenant {}", instanceId, sharingInstance.getTargetTenantId());
+      getInstanceById(instanceId, sharingInstance.getTargetTenantId(), targetInstanceCollection)
+        .onFailure(failure -> {
+          if (failure.getClass().equals(NotFoundException.class)) {
+            LOGGER.info("handle :: instance {} not found on target tenant: {}", instanceId, sharingInstance.getTargetTenantId());
+            getInstanceById(instanceId, sharingInstance.getSourceTenantId(), sourceInstanceCollection)
+              .onSuccess(srcInstance -> {
+                if (srcInstance == null) {
+                  String errorMessage = format("instance %s not found on source tenant: %s", instanceId, sharingInstance.getSourceTenantId());
+                  sendCompleteEventToKafka(tenantId, sharingInstance, kafkaConfig, record.headers());
```

Comment:
I can't tell if the complete event will have an error within...

Reply:
Sure. This message should inform the user that the instance in the source does not exist.

Reply:
Fixed. Copy-Past error. Thanks.